base/sort.jl

```diff
@@ -1112,6 +1112,13 @@ function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsig
         shift < bits || return true
     end
 end
+
+macro repeat4x(a)
+    quote
+        $a; $a; $a; $a
+    end |> esc
+end
```

Comment:
Would `Cartesian.@nexprs 4` be enough instead of defining a specialised macro?

Reply:
Cartesian.@nexprs would work. I am not 100% happy with the requirement of nexprs to require an anonymous function. But I am not set on a specific implementation. Maybe there is an even better way to do it?

Here are some implementation possibilities to compare them directly:
```
while k <= hi - 4
    @repeat4x begin
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
    end
end
```

```
while k <= hi - 4
    Base.Cartesian.@nexprs 4 _ -> begin
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
    end
end
```

```
while k <= hi - 4
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1 
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1               
end
```

Reply:
Resolved by [cb97f9f](https://github.com/JuliaLang/julia/pull/48459/commits/cb97f9fd125fa45de56e775c5bff179765a6027f).

Reply:
I guess `@nexprs` could be special cased to allow a raw `begin` block for a 0-arg function as well. I have some plans for making the whole macro family of `Cartesian` a bit more useful/more user friendly, so I'll keep that in mind.